xdsl/dialects/pdl.py

```diff
@@ -0,0 +1,323 @@
+        return ApplyNativeConstraintOp.build(
+            result_types=[],
+            operands=[args],
+            attributes={"name": StringAttr(name)})
+
+
+@irdl_op_definition
+class ApplyNativeRewriteOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlapply_native_rewrite-mlirpdlapplynativerewriteop
+    """
+    name: str = "pdl.apply_native_rewrite"
+    # https://github.com/xdslproject/xdsl/issues/98
+    # name: OpAttr[StringAttr]
+    args: Annotated[VarOperand, AnyPDLType]
+    res: Annotated[VarOpResult, AnyPDLType]
+
+    def verify_(self) -> None:
+        if 'name' not in self.attributes:
+            raise Exception("ApplyNativeRewriteOp requires a 'name' attribute")
+
+        if not isinstance(self.attributes['name'], StringAttr):
+            raise Exception("expected 'name' attribute to be a StringAttr")
+
+    @staticmethod
+    def get(name: str, args: list[SSAValue],
+            result_types: list[Attribute]) -> ApplyNativeRewriteOp:
+
+        return ApplyNativeRewriteOp.build(
+            result_types=[result_types],
+            operands=[args],
+            attributes={"name": StringAttr(name)})
+
+
+@irdl_op_definition
+class AttributeOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlattribute-mlirpdlattributeop
+    """
+    name: str = "pdl.attribute"
+    value: OptOpAttr[Attribute]
+    value_type: Annotated[OptOperand, TypeType]
+    output: Annotated[OpResult, AttributeType]
+
+
+@irdl_op_definition
+class EraseOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlerase-mlirpdleraseop
+    """
+    name: str = "pdl.erase"
+    op_value: Annotated[Operand, OperationType]
+
+
+@irdl_op_definition
+class OperandOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdloperand-mlirpdloperandop
+    """
+    name: str = "pdl.operand"
+    value_type: Annotated[OptOperand, TypeType]
+    value: Annotated[OpResult, ValueType]
+
+
+@irdl_op_definition
+class OperandsOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdloperands-mlirpdloperandsop
+    """
+    name: str = "pdl.operands"
+    value_type: Annotated[Operand, RangeType[TypeType]]
+    value: Annotated[OpResult, RangeType[ValueType]]
+
+
+@irdl_op_definition
+class OperationOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdloperation-mlirpdloperationop
+    """
+    name: str = "pdl.operation"
+    opName: OptOpAttr[StringAttr]
+    attributeValueNames: OpAttr[ArrayAttr[StringAttr]]
+
+    operandValues: Annotated[VarOperand, ValueType | RangeType[ValueType]]
+    # in PDL docs, this is just a handle to AttributeType, not a range.
+    attributeValues: Annotated[VarOperand,
+                               AttributeType | RangeType[AttributeType]]
+    typeValues: Annotated[VarOperand, TypeType | RangeType[TypeType]]
+    op: Annotated[OpResult, OperationType]
```

Comment:
Let's also make the casing consistent in this file. Seems like some camelcase sneaked into my commits

Reply:
There's a bit of weirdness because the attribute names are coupled to MLIR spelling, and the results/operands are not, so suddenly they become inconsistent with each other. I'd prefer to keep it consistent within a class rather than partially with the rest of the project, and partially with MLIR

Reply:
@math-fehr, @webmiche have you encountered this before, any thoughts?

Reply:
I just realized that even in MLIR this is not consistent. In PDL they use camelCase, for e.g. func.func they use snake_case, see [here](https://mlir.llvm.org/docs/Dialects/Func/#funcfunc-mlirfuncfuncop)

I am fine with either as long as we keep it consistent

Reply:
So my experience is that this is all over the place everywhere in MLIR, that already constrains us quite a bit (i.e. we can't reeally decide on our own). In general, I tried to use snake_case for fields of ops and camelcase for op names. But I don't really care about this a lot, to be entirely honest.

Reply:
I would use `came_case` everywhere we can, so everywhere besides where MLIR decided to be inconsistent!

Reply:
Just to confirm, are you sure this is the case in the generated code?
Sometimes TableGen generates `snake_case` from the `CamelCase` names

Reply:
My intuition is to go with the MLIR names exactly, especially if we're going to generate the definitions automatically from TableGen one day.